types/list.go

```diff
@@ -15,10 +15,22 @@ import (
 )
 
 var (
-	_ attr.Type  = ListType{}
-	_ attr.Value = &List{}
+	_ ListTypable  = ListType{}
+	_ ListValuable = &List{}
 )
 
+type ListTypable interface {
+	attr.Type
+
+	ValueFromList(context.Context, List) (attr.Value, diag.Diagnostics)
```

Comment:
Another thing we should likely consider for these is that the returned attr.Value can be constrained further to only allowing the value interface specific to the type.

```
// ValueFromList should return the equivalent value from a given List.
ValueFromList(context.Context, List) (ListValuable, diag.Diagnostics)
```